File: src/nasdaq/decode_fast_log_extended.py
import struct
import csv
from datetime import datetime

# Message Types Mapping
MESSAGE_TYPES = {
    1: "ORDER_ADD",
    2: "ORDER_EXEC",
    3: "ORDER_CANCEL",
    4: "DEPTH_UPDATE",
    5: "MARKET_STATUS",
    6: "SPECIAL_NEWS"
}

def decode_log_file(input_file="fast_extended_10k.log", output_csv="decoded_fast_log.csv"):
    rows = []

    with open(input_file, "rb") as f:
        while True:
            # Read message type (1 byte) + timestamp (8 bytes)
            header = f.read(9)
            if not header or len(header) < 9:
                break

            msg_type, ts = struct.unpack("Bd", header)
            timestamp = datetime.fromtimestamp(ts).isoformat()
            msg_type_str = MESSAGE_TYPES.get(msg_type, "UNKNOWN")

            # Decode based on message type
            if msg_type == 1:  # ORDER_ADD
                body = f.read(12)
                if len(body) < 12: break
                order_id, volume, price = struct.unpack("IIf", body)
                print(f"{timestamp} | ORDER_ADD: ID={order_id}, Vol={volume}, Price={price}")
                rows.append([timestamp, msg_type_str, order_id, volume, price, "", "", "", ""])
            
            elif msg_type == 2:  # ORDER_EXEC
                body = f.read(8)
                if len(body) < 8: break
                order_id, exec_qty = struct.unpack("II", body)
                print(f"{timestamp} | ORDER_EXEC: ID={order_id}, Qty={exec_qty}")
                rows.append([timestamp, msg_type_str, order_id, "", "", exec_qty, "", "", ""])
            
            elif msg_type == 3:  # ORDER_CANCEL
                body = f.read(4)
                if len(body) < 4: break
                order_id, = struct.unpack("I", body)
                print(f"{timestamp} | ORDER_CANCEL: ID={order_id}")
                rows.append([timestamp, msg_type_str, order_id, "", "", "", "", "", ""])
            
            elif msg_type == 4:  # DEPTH_UPDATE
                body = f.read(12)
                if len(body) < 12: break
                level, bid, ask = struct.unpack("Iff", body)
                print(f"{timestamp} | DEPTH_UPDATE: Level={level}, Bid={bid}, Ask={ask}")
                rows.append([timestamp, msg_type_str, "", "", "", "", level, bid, ask])
            
            elif msg_type == 5:  # MARKET_STATUS
                body = f.read(1)
                if len(body) < 1: break
                status, = struct.unpack("B", body)
                status_str = ["CLOSED", "OPEN", "HALT"][status] if status < 3 else "UNKNOWN"
                print(f"{timestamp} | MARKET_STATUS: {status_str}")
                rows.append([timestamp, msg_type_str, "", "", "", "", "", "", status_str])
            
            elif msg_type == 6:  # SPECIAL_NEWS
                body = f.read(50)
                if len(body) < 50: break
                headline = body.rstrip(b'\x00').decode("utf-8", errors="ignore")
                print(f"{timestamp} | SPECIAL_NEWS: {headline}")
                rows.append([timestamp, msg_type_str, "", "", "", "", "", "", headline])

    # Write CSV
    with open(output_csv, "w", newline='') as csvfile:
        writer = csv.writer(csvfile)
        writer.writerow([
            "Timestamp", "MessageType", "OrderID", "Volume", "Price",
            "ExecQty", "DepthLevel", "Bid", "Ask/Status/Headline"
        ])
        writer.writerows(rows)
    
    print(f"\n[✓] Decoded messages saved to {output_csv}")

if __name__ == "__main__":
    decode_log_file()

File: src/nasdaq/decode_fast_log_extended.rs
use std::fs::File;
use std::io::{BufReader, Read};
use std::error::Error;
use chrono::{NaiveDateTime, Utc, DateTime};
use csv::WriterBuilder;

const MESSAGE_TYPES: [&str; 7] = [
    "UNKNOWN", "ORDER_ADD", "ORDER_EXEC", "ORDER_CANCEL",
    "DEPTH_UPDATE", "MARKET_STATUS", "SPECIAL_NEWS",
];

fn decode_log_file(input_file: &str, output_csv: &str) -> Result<(), Box<dyn Error>> {
    let file = File::open(input_file)?;
    let mut reader = BufReader::new(file);

    let mut wtr = WriterBuilder::new().has_headers(true).from_path(output_csv)?;
    wtr.write_record(&[
        "Timestamp", "MessageType", "OrderID", "Volume", "Price",
        "ExecQty", "DepthLevel", "Bid", "Ask/Status/Headline"
    ])?;

    loop {
        let mut header = [0u8; 9];
        if reader.read_exact(&mut header).is_err() {
            break;
        }
        let msg_type = header[0];
        let ts = f64::from_le_bytes(header[1..9].try_into().unwrap());
        let secs = ts as i64;
        let nsecs = ((ts.fract() * 1e9) as u32);
        let timestamp = DateTime::<Utc>::from_utc(
            NaiveDateTime::from_timestamp_opt(secs, nsecs).unwrap(),
            Utc
        ).to_rfc3339();

        let msg_type_str = MESSAGE_TYPES.get(msg_type as usize).unwrap_or(&"UNKNOWN");

        match msg_type {
            1 => { // ORDER_ADD
                let mut body = [0u8; 12];
                if reader.read_exact(&mut body).is_err() { break; }
                let order_id = u32::from_le_bytes(body[0..4].try_into().unwrap());
                let volume = u32::from_le_bytes(body[4..8].try_into().unwrap());
                let price = f32::from_le_bytes(body[8..12].try_into().unwrap());
                println!("{timestamp} | ORDER_ADD: ID={order_id}, Vol={volume}, Price={price}");
                wtr.write_record(&[
                    &timestamp, msg_type_str, &order_id.to_string(),
                    &volume.to_string(), &price.to_string(),
                    "", "", "", ""
                ])?;
            },
            2 => { // ORDER_EXEC
                let mut body = [0u8; 8];
                if reader.read_exact(&mut body).is_err() { break; }
                let order_id = u32::from_le_bytes(body[0..4].try_into().unwrap());
                let exec_qty = u32::from_le_bytes(body[4..8].try_into().unwrap());
                println!("{timestamp} | ORDER_EXEC: ID={order_id}, Qty={exec_qty}");
                wtr.write_record(&[
                    &timestamp, msg_type_str, &order_id.to_string(),
                    "", "", &exec_qty.to_string(), "", "", ""
                ])?;
            },
            3 => { // ORDER_CANCEL
                let mut body = [0u8; 4];
                if reader.read_exact(&mut body).is_err() { break; }
                let order_id = u32::from_le_bytes(body.try_into().unwrap());
                println!("{timestamp} | ORDER_CANCEL: ID={order_id}");
                wtr.write_record(&[
                    &timestamp, msg_type_str, &order_id.to_string(),
                    "", "", "", "", "", ""
                ])?;
            },
            4 => { // DEPTH_UPDATE
                let mut body = [0u8; 12];
                if reader.read_exact(&mut body).is_err() { break; }
                let level = u32::from_le_bytes(body[0..4].try_into().unwrap());
                let bid = f32::from_le_bytes(body[4..8].try_into().unwrap());
                let ask = f32::from_le_bytes(body[8..12].try_into().unwrap());
                println!("{timestamp} | DEPTH_UPDATE: Level={level}, Bid={bid}, Ask={ask}");
                wtr.write_record(&[
                    &timestamp, msg_type_str, "", "", "",
                    "", &level.to_string(), &bid.to_string(), &ask.to_string()
                ])?;
            },
            5 => { // MARKET_STATUS
                let mut body = [0u8; 1];
                if reader.read_exact(&mut body).is_err() { break; }
                let status = body[0] as usize;
                let status_str = ["CLOSED", "OPEN", "HALT"]
                    .get(status).unwrap_or(&"UNKNOWN");
                println!("{timestamp} | MARKET_STATUS: {status_str}");
                wtr.write_record(&[
                    &timestamp, msg_type_str, "", "", "",
                    "", "", "", status_str
                ])?;
            },
            6 => { // SPECIAL_NEWS
                let mut body = [0u8; 50];
                if reader.read_exact(&mut body).is_err() { break; }
                let headline = String::from_utf8(
                    body.iter().cloned().take_while(|&b| b != 0).collect()
                ).unwrap_or_default();
                println!("{timestamp} | SPECIAL_NEWS: {headline}");
                wtr.write_record(&[
                    &timestamp, msg_type_str, "", "", "",
                    "", "", "", &headline
                ])?;
            },
            _ => break,
        }
    }

    wtr.flush()?;
    println!("\n[✓] Decoded messages saved to {output_csv}");
    Ok(())
}

fn main() -> Result<(), Box<dyn Error>> {
    // Change these file names as needed
    decode_log_file("fast_extended_10k.log", "decoded_fast_log.csv")
}
